Source/UI/Activation/Scenes/AuthenticationOTP/AuthenticationOTPViewController.swift

```diff
@@ -0,0 +1,190 @@
+            alert.addAction(UIAlertAction(title: strings.notValidAlertButton, style: .default, handler: nil))
+            present(alert, animated: true, completion: nil)
+            return
+        }
+        router?.routeToKeyExchange(otp: otp)
+    }
+    
+    // MARK: - Keyboard handling
+    
+    var keyboardObserver: Any?
+    
+    private func registerForKeyboardNotifications() {
+        if keyboardObserver != nil {
+            return
+        }
+        keyboardObserver = NotificationCenter.default.addObserver(forName: UIResponder.keyboardWillShowNotification, object: nil, queue: .main) { [weak self] notification in
+            guard let `self` = self else {
+                return
+            }
+            if let keyboardFrame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue {
+                let keyboardRectangle = keyboardFrame.cgRectValue
+                let keyboardHeight = keyboardRectangle.height - self.bottomSafeArea + 20.0
+                self.bottomKeyboardConstraint?.constant = keyboardHeight
+            }
+        }
+    }
+
+    private func unregisterForKeyboardNotifications() {
+        if let observer = keyboardObserver {
+            NotificationCenter.default.removeObserver(observer)
+            keyboardObserver = nil
+        }
+    }
+    
+    // MARK: - Presentation
+    
+    @IBOutlet private weak var hintLabel: UILabel!
+    @IBOutlet private weak var cancelButtonItem: UIBarButtonItem!
+    @IBOutlet private weak var confirmButton: UIButton?
+    @IBOutlet private weak var otpInput: UITextField!
+    
+    @IBOutlet private weak var bottomKeyboardConstraint: NSLayoutConstraint?
+    
+    var bottomSafeArea: CGFloat = 0.0
+
+    var textFieldDefaultColor: UIColor?
+    var textFieldBlinkColor: UIColor?
+    
+    // MARK: -
+    
+    open override func prepareUI() {
+        let uiData = uiDataProvider.uiDataForAuthenticationOTP
+        let commonStrings = uiDataProvider.uiCommonStrings
+        let theme = uiDataProvider.uiTheme
+        
+        // Apply texts & images
+        self.title = uiData.strings.sceneTitle
+        hintLabel?.text = uiData.strings.sceneDescription
+        confirmButton?.setTitle(uiData.strings.confirmButton, for: .normal)
+        cancelButtonItem.title = commonStrings.cancelTitle
+        otpInput.accessibilityLabel = uiData.strings.accessibilityInput
+
+        // Apply themes
+        
+        // background should be clear color - no image
+        configureBackground(image: nil, color: theme.common.backgroundColor)
+        hintLabel?.textColor = theme.common.textColor
+        confirmButton?.applyButtonStyle(theme.buttons.primary)
+        otpInput.applyTextFieldStyle(theme.authOTP.otpCode)
+        
+    }
+    
+    @objc private func otpFieldDidChange() {
+        confirmButton?.isEnabled = !(otpInput.text?.isEmpty ?? true)
```

Comment:
Oh, one more thing. We should handle the text input via the `UITextFieldDelegate` and do not allow to enter too long strings. We should set a maximum allowed OTP length to some sane value, like 128 or similar. This is because a penetration testers typically do their exercise and put absurdly long strings into the app and wait for a crash :)

Reply:
Yep, good idea :) Will do